app/components/file/browse/file-explorer/file-explorer-workspace.ts

```diff
@@ -0,0 +1,161 @@
+import { log } from "app/utils";
+import { BehaviorSubject, Observable } from "rxjs";
+
+export interface FileSource {
+    name?: string;
+    navigator: FileNavigator;
+    openedFiles?: string[];
+}
+
+export interface OpenedFile {
+    path: string;
+    source: FileSource;
+    loader: FileLoader;
+}
+function sourcesFrom(data: FileNavigator | FileSource | FileSource[]): FileSource[] {
+    if (data instanceof FileNavigator) {
+        return [{
+            name: "Files",
+            navigator: data,
+        }];
+    } else if (Array.isArray(data)) {
+        return data;
+    } else {
+        return [data];
+    }
+}
+
+export interface CurrentNode {
+    source: FileSource;
+    path: string;
+    treeNode: FileTreeNode;
+}
+
+export class FileExplorerWorkspace {
+    public sources: FileSource[];
+    public openedFiles: Observable<OpenedFile[]>;
+    // public openedFolder: Observable<any>;
+    public currentSource: Observable<FileSource>;
+    public currentNode: Observable<CurrentNode>;
+
+    private _currentSource = new BehaviorSubject<FileSource>(null);
+    private _currentPath = new BehaviorSubject("");
+    private _openedFiles = new BehaviorSubject<OpenedFile[]>([]);
+    // private _openedFolder = new BehaviorSubject<any>(null);
+
+    constructor(data: FileNavigator | FileSource | FileSource[]) {
+        this.sources = sourcesFrom(data);
+        this.openedFiles = this._openedFiles.asObservable();
+        // this.openedFolder = this._openedFolder.asObservable();
+        this.currentSource = this._currentSource.asObservable();
+        this._currentSource.next(this.sources.first());
+
+        this.currentNode = this._currentSource.flatMap((source) => {
+            return Observable.combineLatest(this._currentPath, source.navigator.tree).map(([path, tree]) => {
+                const node = tree.getNode(path);
+                return {
+                    source,
+                    path: node.path,
+                    treeNode: node,
+                };
+            });
+        }).shareReplay(1);
+
+        this._openInitalFiles();
+    }
+
+    public navigateTo(path: string, source?: FileSource) {
+        source = this._getSource(source);
+        if (this._currentPath.value === path && this._currentSource.value === source) { return; }
+
+        this._currentPath.next(path);
+        this._currentSource.next(source);
+        source.navigator.getNode(path).subscribe((node) => {
+            if (!node) { return; }
+            if (node.isDirectory) {
+                source.navigator.loadPath(path);
+            } else {
+                this.openFile(node.path, source);
+            }
+        });
+    }
+
+    /**
+     * Go back up one level
+     */
+    public goBack() {
+        const path = this._currentPath.value;
+        if (path === "") { return; }
+        this.navigateTo(path.split("/").slice(0, -1).join("/"), this._currentSource.value);
```

Comment:
you have used ".split(/[\\\/]/)" in the past, do we need to do that here to take into account / or \

Reply:
the path is already suppose to be normalized in the tree though

Reply:
but there is a util method dirname I made I should use here instead